### PR TITLE
[Rewards 3.0] Add connect account UI

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/39876
Resolves https://github.com/brave/brave-browser/issues/39852

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Note: QA may want to wait to test this until https://github.com/brave/brave-browser/issues/39877 is also complete.

* Enable the Rewards 3.0 flag: `#brave-rewards-new-rewards-ui` (Show the new Rewards UI)
* With a new profile, enable Rewards.
* On either the rewards page or rewards panel, press the "Connect account" button.
  * (If pressed on the panel, it should open up the Rewards page.)
* Verify that connect account screen is displayed with the correct options.
* Select a provider and complete the provider login flow.
* Ensure that user is in the connected state after login flow is completed.

### CODE REVIEW

File: browser/brave_rewards/test/rewards_page_browsertest.cc [hunk 1]
```diff
@@ -120,6 +171,43 @@ class RewardsPageBrowserTest : public InProcessBrowserTest {
     LoadScript("test_setup.js");
   }
 
+  using RequestHandlerResult = std::optional<std::pair<int, std::string>>;
+  using RequestHandler =
+      base::RepeatingCallback<RequestHandlerResult(const GURL&,
+                                                   const std::string&)>;
+
+  void SetRequestHandler(RequestHandler handler) {
```
Comment (zenparsing) [inline browser/brave_rewards/test/rewards_page_browsertest.cc:179]: Request handling has been moved out of the test scripts and into the tests on the C++ side.

File: components/brave_rewards/browser/rewards_service.h [hunk 2]
```diff
@@ -275,6 +275,13 @@ class RewardsService : public KeyedService {
                                      const std::string& query,
                                      ConnectExternalWalletCallback) = 0;
 
+  // Completes an external wallet login flow for the specified wallet provider
+  // using a collection of parameters returned from the provider's login page.
+  virtual void ConnectExternalWallet(
```
Comment (zenparsing) [inline components/brave_rewards/browser/rewards_service.h:280]: For the 3.0 Rewards page, the parsing of the URL is performed on the front-end. This new overload allows the front-end to submit the parsed info to the engine, bypassing URL parsing.

File: components/resources/brave_components_strings.grd [hunk 3]
```diff
@@ -550,50 +550,8 @@
       <message name="IDS_BRAVE_REWARDS_LOCAL_ADS_CONFIRMATION_TYPE_VIEW" desc="">Viewed</message>
 
       <message name="IDS_BRAVE_REWARDS_LOCAL_REDIRECT_MODAL_CLOSE" desc="">Close</message>
-      <message name="IDS_BRAVE_REWARDS_LOCAL_REDIRECT_MODAL_DEVICE_LIMIT_REACHED_TEXT" desc="">Your Brave Rewards cannot be verified because you've reached the maximum verified device limit.</message>
```
Comment (zenparsing) [inline components/resources/brave_components_strings.grd:553]: These strings have been moved into `rewards_strings.grdp` and renamed to match the key used on the front-end.

File: components/brave_rewards/resources/rewards_page/components/connect_account.tsx [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  getExternalWalletProviderName,
+  isExternalWalletProviderAllowed
+} from '../../shared/lib/external_wallet'
+
+import * as routes from '../lib/app_routes'
+import * as urls from '../../shared/lib/rewards_urls'
+
+import { style } from './connect_account.style'
+
+type ConnectState = 'loading' | 'error' | ''
```
Comment (emerick) [inline components/brave_rewards/resources/rewards_page/components/connect_account.tsx:32]: Does the empty state indicate that the wallet is connected? Or is it ambiguous?

Comment (zenparsing) [reply components/brave_rewards/resources/rewards_page/components/connect_account.tsx:32]: This is just the form state - whether to show the spinner or an error message or the normal state. The naming is confusing though; I'll find a better name.

Comment (emerick) [reply components/brave_rewards/resources/rewards_page/components/connect_account.tsx:32]: Gotcha! Maybe `LoadingState`?

**Review by emerick — APPROVED** [review]

File: components/resources/rewards_strings.grdp [hunk 5]
```diff
@@ -550,4 +550,140 @@
...
+    Custodial accounts are accounts with trusted third-party services that hold your BAT for you.
+  </message>
+  <message name="IDS_REWARDS_CONNECT_SELF_CUSTODY_TITLE" desc="">
+    Self-custody
+  </message>
+  <message name="IDS_REWARDS_CONNECT_SELF_CUSTODY_TOOLTIP" desc="">
+    Self-custody gives you exclusive control over your BAT, along with complete responsibility for its security.
+  </message>
+  <message name="IDS_REWARDS_CONNECT_SELF_CUSTODY_NOTE" desc="">
+    By choosing a self-custody wallet, Rewards will be paid out as SPL-BAT on the Solana network.
```
Comment (mkarolin) [inline components/resources/rewards_strings.grdp:567]: I know these are just moved, but this sounds a bit odd. Maybe `When a self-custody wallet is chosen,...`

File: components/resources/rewards_strings.grdp [hunk 6]
```diff
@@ -550,4 +550,140 @@
...
+    Currently not available in your region
+  </message>
+  <message name="IDS_REWARDS_CONNECT_TEXT" desc="">
+    To begin earning Basic Attention Token (BAT) rewards, connect an account. BAT you earn will be automatically transferred to your account every month.
+  </message>
+  <message name="IDS_REWARDS_CONNECT_SOLANA_BUTTON_LABEL" desc="">
+    Brave Wallet
+  </message>
+  <message name="IDS_REWARDS_CONNECT_SOLANA_MESSAGE" desc="">
+    Or other wallet that supports Solana
```
Comment (mkarolin) [inline components/resources/rewards_strings.grdp:585]: another?

File: components/resources/rewards_strings.grdp [hunk 7]
```diff
@@ -550,4 +550,140 @@
...
+    Different account
+  </message>
+  <message name="IDS_REWARDS_AUTHORIZE_PROVIDER_UNAVAILABLE_TEXT_1" desc="">
+    It looks like connecting <ph name="PROVIDER">$1<ex>Uphold</ex></ph> to Brave Rewards is temporarily unavailable. Please try again later.
+  </message>
+  <message name="IDS_REWARDS_AUTHORIZE_PROVIDER_UNAVAILABLE_TEXT_2" desc="">
+    You can view <ph name="BEGIN_LINK">$2</ph>service status here<ph name="END_LINK">$3</ph>.
+  </message>
+  <message name="IDS_REWARDS_AUTHORIZE_PROVIDER_UNAVAILABLE_TITLE" desc="">
+    Connecting to selected custodian temporarily unavailable
```
Comment (mkarolin) [inline components/resources/rewards_strings.grdp:650]: Connecting to selected custodian **is** temporarily unavailable

File: components/resources/rewards_strings.grdp [hunk 8]
```diff
@@ -550,4 +550,140 @@
...
+    Error: BAT unavailable
+  </message>
+  <message name="IDS_REWARDS_AUTHORIZE_UPHOLD_INSUFFICIENT_CAPABILITIES_TEXT" desc="">
+    According to Uphold, there are currently some limitations on your Uphold account. Please log in to your Uphold account and check whether there are any notices or remaining account requirements to complete, then try again. For example, you may need to submit more information to Uphold.
+  </message>
+  <message name="IDS_REWARDS_AUTHORIZE_UPHOLD_INSUFFICIENT_CAPABILITIES_TITLE" desc="">
+    Limited Uphold account functionality
+  </message>
+  <message name="IDS_REWARDS_AUTHORIZE_SIGNATURE_VERIFICATION_ERROR_TEXT" desc="">
+    There was a problem validating your browser's request to connect Brave Rewards to your <ph name="PROVIDER">$1<ex>Uphold</ex></ph> account. You can <ph name="BEGIN_LINK">$2</ph>contact support<ph name="END_LINK">$3</ph> for more help.
```
Comment (mkarolin) [inline components/resources/rewards_strings.grdp:674]: Do we need `browser's` here? It makes it sound like "your browser" isn't Brave. Or maybe say "the browser's request"?

**Review by mkarolin — APPROVED** [review]
++ with nits

File: components/resources/rewards_strings.grdp (hunk 5 above)
Comment (Miyayes) [reply components/resources/rewards_strings.grdp:567]: Let's use:
>With self-custody, your BAT earnings will be paid out as SPL-BAT on the Solana network.

File: components/resources/rewards_strings.grdp (hunk 6 above)
Comment (Miyayes) [reply components/resources/rewards_strings.grdp:585]: Keep as is

File: components/resources/rewards_strings.grdp (hunk 7 above)
Comment (Miyayes) [reply components/resources/rewards_strings.grdp:650]: Keep as is

File: components/resources/rewards_strings.grdp (hunk 8 above)
Comment (Miyayes) [reply components/resources/rewards_strings.grdp:674]: Good suggestion. Change to the following (note there are other changes as well): 

>There was a problem validating your request to connect your <ph name="PROVIDER">$1<ex>Uphold</ex></ph> account to Brave Rewards. You can <ph name="BEGIN_LINK">$2</ph>contact support<ph name="END_LINK">$3</ph> for more help.

File: components/resources/brave_components_strings.grd (hunk 3 above)
Comment (Miyayes) [reply components/resources/brave_components_strings.grd:553]: Wherever it was moved, we should update this to say:

>You've reached the maximum device limit for this account.

**Pull request merged**